Tools/scripts/run_coverage.py

```diff
@@ -0,0 +1,275 @@
+#!/usr/bin/env python
+
+"""
+Runs tests with gcov coverage support.
+
+ AP_FLAKE8_CLEAN
+"""
+import argparse
+import os
+import shutil
+import subprocess
+import sys
+
+os.environ['PYTHONUNBUFFERED'] = '1'
+
+tools_dir = os.path.dirname(os.path.realpath(__file__))
+root_dir = os.path.realpath(os.path.join(tools_dir, '../..'))
+
+
+class CoverageRunner(object):
+    """Coverage Runner Class."""
+
+    def __init__(self):
+        """Init method.
+
+        Set the files Path.
+        """
+        self.REPORT_DIR = os.path.join(root_dir, "reports/lcov-report")
+        self.INFO_FILE = os.path.join(root_dir, self.REPORT_DIR, "lcov.info")
+        self.INFO_FILE_BASE = os.path.join(root_dir, self.REPORT_DIR, "lcov_base.info")
+        self.LCOV_LOG = os.path.join(root_dir, "GCOV_lcov.log")
+        self.GENHTML_LOG = os.path.join(root_dir, "GCOV_genhtml.log")
+        self.CI = os.environ.get("CI", False)
+
+    def progress(self, string):
+        """Pretty printer."""
+        print("****** %s" % (string,))
+
+    def init_coverage(self):
+        """Initialize ArduPilot for coverage.
+
+        This need to be run with the binaries builded.
+        """
+        self.progress("Initilizing Coverage...")
+        self.progress("Removing previous reports")
+        try:
+            shutil.rmtree(self.REPORT_DIR)
+        except OSError as e:
+            print("Error: %s : %s" % (self.REPORT_DIR, e.strerror))
+
+        try:
+            os.makedirs(self.REPORT_DIR)
+        except OSError as error:
+            print(error)
+
+        self.progress("Checking that vehicles binaries are set up and built")
+        examples_dir = os.path.join(root_dir, 'build/linux/examples')
+        binaries_dir = os.path.join(root_dir, 'build/sitl/bin')
+        is_build_correct = self.check_build("example", examples_dir) and self.check_build("binaries", binaries_dir)
+        if not is_build_correct:
+            self.run_build()
+
+        self.progress("Zeroing previous build")
+        retcode = subprocess.call(["lcov", "--zerocounters", "--directory", root_dir])
+        if retcode != 0:
+            self.progress("Failed with retcode (%s)" % retcode)
```

Comment:
Why are you continuing to execute this script when things are failing?

When you're mucking around with operations like `rmtree`, you *quit* when things aren't working right!

Reply:
yep, I just miss those. corrected now